Fix missing icudata library and runtime error

asdf288 opened the pull request:
The previous icu package built an empty libicudata.so library and was missing a patch from upstream LibreELEC that fixes a runtime error on some platforms. This should fix #1189.

ToKe79 commented:
Hi,
thank you for the PR - can you list the files/folders that need to be removed in $SYSROOT_PREFIX and $TOOLCHAIN in order to correctly rebuild icu:host and icu:target?
It was in the original PR #1219, but that one currently does not contain any commits.
I will remove those files/folders manually in our build folders.
Thank you!

asdf288 commented:
@ToKe79 To clarify what the problem is: If this change is pulled and an existing build of Lakka is upgraded, it will fail because ICU tries to link against its old toolchain library instead of the newly built libraries (this is because the toolchain comes first in the library search order). This is why I added code to clean the toolchain before.

For Lakka's nightly build system, this shouldn't be a problem because it seems to delete the build folder and start from scratch every time.

In the end, I removed the code again because Lakka doesn't need it and no other packages do this. It's just a convenience for users who rebuild their existing images.

The files should be:
$TOOLCHAIN/lib/libicu*
$TOOLCHAIN/include/unicode

$SYSROOT_PREFIX/usr/include/unicode
$SYSROOT_PREFIX/usr/lib/libicu*

ToKe79 commented:
No, we keep the build folders for nightly/PR builds, otherwise we have to rebuild 20 targets, which takes some time.
I will wait until the job is finished and then remove the files from build folder and start a new job.

ToKe79 commented:
Hi,
this PR (4383f2deb51e7490ff878be80d06dd3492551d2d) did not build correctly for any target - all (20) builds failed.
Logs can be found here: http://nightly.builds.lakka.tv/PRs/pr_1221_000236/

_This is an automated message_

ToKe79 commented:
Hi,
this PR (4383f2deb51e7490ff878be80d06dd3492551d2d) did not build correctly for all targets.
Following targets failed (1):
- RK3399.arm

Logs and images for testing can be found here: http://nightly.builds.lakka.tv/PRs/pr_1221_000238/

_This is an automated message_

ToKe79 commented:
@asdf288 can you test if your changes got applied? please use an image from [here](http://nightly.builds.lakka.tv/PRs/pr_1221_000238/).

asdf288 commented:
@ToKe79 The libs are OK now, but I get:

[ERROR] Failed to open libretro core: "/tmp/cores/easyrpg_libretro.so"
[ERROR] Error(s): libicui18n.so.61: cannot open shared object file: No such file or directory

So it's still linked against ICU 61, you need to rebuild liblcf (and easyrpg too, just to be safe).

ToKe79 commented:
Thank you for the test. I cleaned the build folders and restarted the build. It will take quite longer now, but new report will appear here as soon as the images are ready.

ToKe79 commented:
And Generic.x86_64 is again out of space... what platform are you testing on?

asdf288 commented:
You mean clean everthing in the build folder? No, that's not necessary! Just rebuild libclf, that should be enough. I have tested it on AMLGX.

ToKe79 commented:
OK, so I cleaned `icu`, `liblcf` and `easyrpg` and restarted the build.

ToKe79 commented:
Hi,
this PR (4383f2deb51e7490ff878be80d06dd3492551d2d) did not build correctly for all targets.
Following targets failed (1):
- OdroidGoAdvance.arm

Logs and images for testing can be found here: http://nightly.builds.lakka.tv/PRs/pr_1221_000239/

_This is an automated message_

ToKe79 commented:
I checked OGA manually, builds OK.
@asdf288 please - retest AMLGX image and confirm, that it works as expected. Thank you!

asdf288 commented:
@ToKe79 It works now! Sorry that it took so much time to install, I hope it wasn't too annoying.

Ghabry commented:
Thanks for solving this!
That patch looks interesting, I will make a copy of it, could be useful for my own ICU cross compiling adventures xD

asdf288 commented:
@Ghabry This solution does not use the smaller custom library, though. It would be cool to have that, but it would require creating another ICU package (one full package for some apps, and the custom one for liblcf), and the custom library would need a nonstandard install location or name suffix, and packages that use it need extra settings to point it to the location. This adds a lot of complexity so I left it out for now, but it's still useful to know about it and might be added in the future, thanks for your help!
